File: styx-js/test/crypto/shamir.test.js
import { describe, test, expect } from '@jest/globals';
import * as fc from 'fast-check';
import {
  ShamirSplitter,
  ShamirReconstructor,
  ShamirShare,
  InsufficientSharesException,
  KeyBackup,
} from '../../src/crypto/shamir.js';
import { IdentityManager } from '../../src/crypto/identity.js';
import { bytesToHex, randomBytes, constantTimeEqual } from '../../src/utils.js';

describe('ShamirSplitter', () => {
  const splitter = new ShamirSplitter();
  const reconstructor = new ShamirReconstructor();

  test('split(secret, 2, 3) returns 3 shares', () => {
    const secret = randomBytes(32);
    const shares = splitter.split(secret, 2, 3);
    expect(shares.length).toBe(3);
    shares.forEach((share) => {
      expect(share).toBeInstanceOf(ShamirShare);
      expect(share.data.length).toBe(32);
    });
  });

  test('shares have indices 1, 2, 3', () => {
    const shares = splitter.split(randomBytes(16), 2, 3);
    expect(shares.map((s) => s.index)).toEqual([1, 2, 3]);
  });

  test('reconstruct with any 2 shares of 3 recovers original', () => {
    const secret = randomBytes(32);
    const shares = splitter.split(secret, 2, 3);

    // Try all 2-combinations: [0,1], [0,2], [1,2]
    const combos = [
      [shares[0], shares[1]],
      [shares[0], shares[2]],
      [shares[1], shares[2]],
    ];

    for (const combo of combos) {
      const recovered = reconstructor.reconstruct(combo);
      expect(bytesToHex(recovered)).toBe(bytesToHex(secret));
    }
  });

  test('reconstruct with 3 of 5 (threshold 3, total 5)', () => {
    const secret = randomBytes(16);
    const shares = splitter.split(secret, 3, 5);
    expect(shares.length).toBe(5);

    // Pick shares [0, 2, 4]
    const recovered = reconstructor.reconstruct([shares[0], shares[2], shares[4]]);
    expect(bytesToHex(recovered)).toBe(bytesToHex(secret));
  });

  test('reconstruct with all shares works', () => {
    const secret = randomBytes(32);
    const shares = splitter.split(secret, 2, 3);
    const recovered = reconstructor.reconstruct(shares);
    expect(bytesToHex(recovered)).toBe(bytesToHex(secret));
  });
});

describe('InsufficientSharesException', () => {
  test('thrown when fewer than 2 shares provided', () => {
    const reconstructor = new ShamirReconstructor();
    const splitter = new ShamirSplitter();
    const shares = splitter.split(randomBytes(16), 2, 3);

    expect(() => reconstructor.reconstruct([shares[0]])).toThrow(
      InsufficientSharesException
    );
    expect(() => reconstructor.reconstruct([])).toThrow(
      InsufficientSharesException
    );
  });
});

describe('ShamirSplitter validation', () => {
  const splitter = new ShamirSplitter();

  test('threshold < 2 throws', () => {
    expect(() => splitter.split(randomBytes(8), 1, 3)).toThrow('Threshold must be at least 2');
  });

  test('totalShares < threshold throws', () => {
    expect(() => splitter.split(randomBytes(8), 3, 2)).toThrow('totalShares must be >= threshold');
  });
});

describe('ShamirShare serialization', () => {
  test('serialize and deserialize roundtrip', () => {
    const share = new ShamirShare(5, randomBytes(32));
    const serialized = share.serialize();
    expect(typeof serialized).toBe('string');

    const deserialized = ShamirShare.deserialize(serialized);
    expect(deserialized.index).toBe(5);
    expect(bytesToHex(deserialized.data)).toBe(bytesToHex(share.data));
  });

  test('multiple shares serialize/deserialize correctly', () => {
    const splitter = new ShamirSplitter();
    const secret = randomBytes(16);
    const shares = splitter.split(secret, 2, 3);

    const serialized = shares.map((s) => s.serialize());
    const deserialized = serialized.map((s) => ShamirShare.deserialize(s));

    const reconstructor = new ShamirReconstructor();
    const recovered = reconstructor.reconstruct(deserialized);
    expect(bytesToHex(recovered)).toBe(bytesToHex(secret));
  });
});

describe('ShamirShare index validation', () => {
  test('index 0 throws', () => {
    expect(() => new ShamirShare(0, randomBytes(8))).toThrow('Share index must be 1-255');
  });

  test('index 256 throws', () => {
    expect(() => new ShamirShare(256, randomBytes(8))).toThrow('Share index must be 1-255');
  });

  test('index 1 is valid', () => {
    expect(() => new ShamirShare(1, randomBytes(8))).not.toThrow();
  });

  test('index 255 is valid', () => {
    expect(() => new ShamirShare(255, randomBytes(8))).not.toThrow();
  });
});

describe('KeyBackup', () => {
  const backup = new KeyBackup();
  const im = new IdentityManager();

  test('backupPrivateKey and restoreFromShares roundtrip', async () => {
    const kp = await im.generate();
    const shares = backup.backupPrivateKey(kp.privateKey, 2, 3);
    expect(shares.length).toBe(3);

    const restored = await backup.restoreFromShares([shares[0], shares[2]], im);
    expect(restored.publicKey.equals(kp.publicKey)).toBe(true);
  });

  test('restoreFromShares with all shares works', async () => {
    const kp = await im.generate();
    const shares = backup.backupPrivateKey(kp.privateKey, 2, 3);
    const restored = await backup.restoreFromShares(shares, im);
    expect(restored.publicKey.equals(kp.publicKey)).toBe(true);
  });
});

describe('property-based test with fast-check', () => {
  const splitter = new ShamirSplitter();
  const reconstructor = new ShamirReconstructor();

  test('random secret with random threshold always reconstructs correctly', () => {
    fc.assert(
      fc.property(
        fc.integer({ min: 1, max: 64 }), // secret length
        fc.integer({ min: 2, max: 5 }),   // threshold
        fc.integer({ min: 0, max: 3 }),   // extra shares beyond threshold
        (secretLen, threshold, extra) => {
          const totalShares = threshold + extra;
          const secret = randomBytes(secretLen);
          const shares = splitter.split(secret, threshold, totalShares);

          // Pick exactly `threshold` shares (first `threshold`)
          const subset = shares.slice(0, threshold);
          const recovered = reconstructor.reconstruct(subset);
          return bytesToHex(recovered) === bytesToHex(secret);
        }
      ),
      { numRuns: 20 }
    );
  });
});

File: packages/crypto_core/test/shamir_test.dart
import 'dart:typed_data';

import 'package:styx_crypto_core/styx_crypto_core.dart';
import 'package:test/test.dart';

void main() {
  final splitter = ShamirSplitter();
  final reconstructor = ShamirReconstructor();

  /// Helper: picks shares at given indices from the list.
  List<ShamirShare> pick(List<ShamirShare> shares, List<int> indices) => [
    for (final i in indices) shares[i],
  ];

  group('Shamir Secret Sharing', () {
    test('T3.19 — 2-of-3 reconstruction (shares 1,2)', () {
      final secret = Uint8List.fromList(List.generate(32, (i) => i));
      final shares = splitter.split(secret: secret);
      final result = reconstructor.reconstruct(pick(shares, [0, 1]));
      expect(result, equals(secret));
    });

    test('T3.20 — 2-of-3 reconstruction (shares 1,3)', () {
      final secret = Uint8List.fromList(List.generate(32, (i) => i));
      final shares = splitter.split(secret: secret);
      final result = reconstructor.reconstruct(pick(shares, [0, 2]));
      expect(result, equals(secret));
    });

    test('T3.21 — 2-of-3 reconstruction (shares 2,3)', () {
      final secret = Uint8List.fromList(List.generate(32, (i) => i));
      final shares = splitter.split(secret: secret);
      final result = reconstructor.reconstruct(pick(shares, [1, 2]));
      expect(result, equals(secret));
    });

    test('T3.22 — 2-of-3 with 1 share: wrong result', () {
      final secret = Uint8List.fromList(List.generate(32, (i) => i));
      final shares = splitter.split(secret: secret);
      // With only 1 share when threshold=2, reconstruction produces
      // a technically valid but incorrect result.
      final result = reconstructor.reconstruct([shares[0]]);
      expect(result, isNot(equals(secret)));
    });

    test('T3.23 — 3-of-5 all C(5,3)=10 combinations', () {
      final secret = Uint8List.fromList(List.generate(32, (i) => i * 3));
      final shares = splitter.split(
        secret: secret,
        threshold: 3,
        totalShares: 5,
      );

      // All 10 combinations of 3 from 5
      final combos = <List<int>>[
        [0, 1, 2],
        [0, 1, 3],
        [0, 1, 4],
        [0, 2, 3],
        [0, 2, 4],
        [0, 3, 4],
        [1, 2, 3],
        [1, 2, 4],
        [1, 3, 4],
        [2, 3, 4],
      ];

      for (final combo in combos) {
        final result = reconstructor.reconstruct(pick(shares, combo));
        expect(
          result,
          equals(secret),
          reason: 'Failed for combo $combo',
        );
      }
    });

    test('T3.24 — 3-of-5 with 2 shares: wrong result', () {
      final secret = Uint8List.fromList(List.generate(32, (i) => i * 3));
      final shares = splitter.split(
        secret: secret,
        threshold: 3,
        totalShares: 5,
      );
      final result = reconstructor.reconstruct(pick(shares, [0, 1]));
      expect(result, isNot(equals(secret)));
    });

    test('T3.25 — 1-byte secret', () {
      final secret = Uint8List.fromList([0x42]);
      final shares = splitter.split(secret: secret);
      final result = reconstructor.reconstruct(pick(shares, [0, 1]));
      expect(result, equals(secret));
    });

    test('T3.26 — 64-byte secret', () {
      final secret = Uint8List.fromList(
        List.generate(64, (i) => (i * 17 + 5) & 0xFF),
      );
      final shares = splitter.split(secret: secret);
      final result = reconstructor.reconstruct(pick(shares, [0, 2]));
      expect(result, equals(secret));
    });

    test('T3.27 — All-zero secret', () {
      final secret = Uint8List(32);
      final shares = splitter.split(secret: secret);
      final result = reconstructor.reconstruct(pick(shares, [0, 1]));
      expect(result, equals(secret));
    });

    test('T3.28 — All-0xFF secret', () {
      final secret = Uint8List.fromList(List.filled(32, 0xFF));
      final shares = splitter.split(secret: secret);
      final result = reconstructor.reconstruct(pick(shares, [1, 2]));
      expect(result, equals(secret));
    });

    test('T3.29 — Share serialize/deserialize round-trip', () {
      final secret = Uint8List.fromList(List.generate(32, (i) => i));
      final shares = splitter.split(secret: secret);

      for (final share in shares) {
        final serialized = share.serialize();
        final deserialized = ShamirShare.deserialize(serialized);
        expect(deserialized.index, equals(share.index));
        expect(deserialized.data, equals(share.data));
      }
    });

    test('T3.30 — Share corrupted: flip 1 byte → wrong reconstruction', () {
      final secret = Uint8List.fromList(List.generate(32, (i) => i));
      final shares = splitter.split(secret: secret);

      // Corrupt share 0 by flipping a byte.
      final corrupted = ShamirShare(
        index: shares[0].index,
        data: Uint8List.fromList(shares[0].data),
      );
      corrupted.data[0] ^= 0xFF;

      final result = reconstructor.reconstruct([corrupted, shares[1]]);
      expect(result, isNot(equals(secret)));
    });
  });

  group('Shamir edge cases', () {
    test('empty secret throws', () {
      expect(
        () => splitter.split(secret: Uint8List(0)),
        throwsArgumentError,
      );
    });

    test('threshold > totalShares throws', () {
      expect(
        () => splitter.split(
          secret: Uint8List.fromList([1]),
          threshold: 3,
          totalShares: 2,
        ),
        throwsArgumentError,
      );
    });

    test('duplicate share indices throws', () {
      final share = ShamirShare(index: 1, data: Uint8List.fromList([42]));
      expect(
        () => reconstructor.reconstruct([share, share]),
        throwsA(isA<InvalidShareException>()),
      );
    });
  });
}
